### ISSUE TITLE
non-coarray vector lhs produces crashes or wrong results with gets

### ISSUE THREAD

**Issue by reinh-bader** [sourceryinstitute/OpenCoarrays#322]
The attached source code was built with
* gfortran 6.3 + opencoarrays 1.6.2
* gfortran 7.0 (preview) + opencoarrays 1.8.2
and produces wrong results when executed with any number of images. The version build with `gfortran -fcoarray=single` executes fine, hence I suspect the problem to be in the run time library. 

The program checks its output. 
For opencoarrays 1.8.2, I get the run time error

```
Fatal error in MPI_Win_unlock_all: Invalid MPI_Win, error stack:
MPI_Win_unlock_all(127): MPI_Win_unlock_all(win=0x55444f4d) failed
MPI_Win_unlock_all(82).: Invalid MPI_Win
Fatal error in MPI_Win_unlock_all: Invalid MPI_Win, error stack:
MPI_Win_unlock_all(127): MPI_Win_unlock_all(win=0x55444f4d) failed
MPI_Win_unlock_all(82).: Invalid MPI_Win
```

[PMR.zip](https://github.com/sourceryinstitute/opencoarrays/files/738796/PMR.zip)

**Comment by zbeekman** [comment]
Can you please include a little bit more information: MPI implementation, system (OS, arch, virtualized?) and how you built and installed OpenCoarrays (with install log if possible)

Thanks!

**Comment by reinh-bader** [comment]
Base MPI is Intel MPI 5.1.3
The system is a X86-based shared memory node with SLES11 SP4 as OS. 
Here is the command line used to build:

CC=gcc FC="gfortran -I/lrz/sys/intel/impi/5.1.3.181/intel64/include/gfortran/7.0.0" cmake /lrz/noarch/src/gcc/opencoarrays/opencoarrays-1.8.2 -DBUILD_SHARED_LIBS:BOOL=ON -DCMAKE_INSTALL_PREFIX=${PWD}

The only log file I found was CMakeFiles/CMakeOutput.log which I have attached (adding a .txt extension, to enable the upload).

The example code was compiled with 
caf vector_subscript_pos_01.f90
and executed with
cafrun -n 2 ./a.out

[CMakeOutput.log.txt](https://github.com/sourceryinstitute/opencoarrays/files/742532/CMakeOutput.log.txt)

Hope this helps.

**Comment by zbeekman** [comment]
Thanks so much @reinh-bader... I think I know of a machine or two that has Intel MPI 5.1.3 if the problem does not show up with MPICH/OpenMPI

**Comment by zbeekman** [comment]
Sorry I have been swamped, and have not had a chance to fully investigate yet, but this is still on my radar

**Comment by reinh-bader** [comment]
No trouble, take your time.

**Comment by zbeekman** [comment]
This is still broken as of 0c2dce7bba41a8cda8ac10c53c456458797d0b11 on 12/26/2017

Here is the test case inlined:

```fortran
program mod_vecsub_01
  implicit none
  integer, parameter :: ndim = 5, vdim = 2
  real :: vec(ndim), res(ndim)[*]
  integer :: idx(vdim)
  integer :: i, me
  logical :: ok[*]

  res = [ (real(i), i=1, ndim) ]
  vec = 0.0
  idx = [ ndim, 1 ]
  ok = .true.
  sync all
  me = this_image()
  vec(idx) = res(1:2)[1]
  if (vec(1) /= 2.0 .or. vec(5) /= 1.0) then
    critical
      ok[1] = .false.
      write(*, *) 'FAIL on image ',me,vec(idx)
    end critical
  end if
  if (me == 1) then
     if (ok) then
       write(*, *) 'OK'
     end if
  end if
end program
```

**Comment by vehre** [comment]
That's got nothing to do with opencoarrays. It's a compiler bug. Gfortran is not respecting the indexed addressing for the lhs *after* the get(). It's trying to do it before.

**Comment by zbeekman** [comment]
@vehre awesome, thanks for the response! @rouson we should look and see if this bug has been filed, and if it has not, make sure we file one and post the tracking number here. Perhaps @afanfa or Sorren might be interested in working on a fix for this?

**Comment by rouson** [comment]
This has been submitted as [Bug 83606] on gfortran with the following slightly simplified code:
```fortran
  integer, parameter :: ndim=5
  integer :: i,vec(ndim)=0, res(ndim)[*]=[ (i, i=1, ndim) ]
  vec([ndim,1]) = res(1:2)[1]
  if (vec(1) /= res(2) .or. vec(ndim) /= res(1)) &
    print *," wrong result ",vec([ndim,1])," on image ",this_image()
end
```
 [Bug 83606]: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=83606

**Comment by zbeekman** [comment]
See also #427.

**Comment by vehre** [comment]
Fixed on all gcc's from 6 on.

**Comment by zbeekman** [comment]
@vehre is the fix in 6.4 and 7.3 or will it be in 6.5 and 7.4?
